src/stress/compute/compute_pass.spec.ts

```diff
@@ -12,26 +13,230 @@ g.test('many')
 GPUComputePipeline.`
   )
-  .unimplemented();
+  .fn(async t => {
+    const kNumElements = 64;
+    const data = new Uint32Array([...iterRange(kNumElements, x => x)]);
+    const buffer = t.makeBufferWithContents(data, GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC);
+    const pipeline = t.device.createComputePipeline({
+      compute: {
+        module: t.device.createShaderModule({
+          code: `
+            [[block]] struct Buffer { data: array<u32>; };
+            [[group(0), binding(0)]] var<storage, read_write> buffer: Buffer;
+            [[stage(compute), workgroup_size(1)]] fn main(
+                [[builtin(global_invocation_id)]] id: vec3<u32>) {
+              buffer.data[id.x] = buffer.data[id.x] + 1u;
+            }
+          `,
+        }),
+        entryPoint: 'main',
+      },
+    });
+    const bindGroup = t.device.createBindGroup({
+      layout: pipeline.getBindGroupLayout(0),
+      entries: [{ binding: 0, resource: { buffer } }],
+    });
+    const kNumIterations = 250_000;
+    for (let i = 0; i < kNumIterations; ++i) {
+      const encoder = t.device.createCommandEncoder();
+      const pass = encoder.beginComputePass();
+      pass.setPipeline(pipeline);
+      pass.setBindGroup(0, bindGroup);
+      pass.dispatch(kNumElements);
+      pass.endPass();
+      t.device.queue.submit([encoder.finish()]);
+    }
+    t.expectGPUBufferValuesEqual(
+      buffer,
+      new Uint32Array([...iterRange(kNumElements, x => x + kNumIterations)])
+    );
+  });
 
 g.test('pipeline_churn')
   .desc(
     `Tests execution of a huge number of compute passes which each use a different
 GPUComputePipeline.`
   )
-  .unimplemented();
+  .fn(async t => {
+    const kNumElements = 64;
+    const data = new Uint32Array([...iterRange(kNumElements, x => x)]);
+    const buffer = t.makeBufferWithContents(data, GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC);
+    const module1 = t.device.createShaderModule({
+      code: `
+        [[block]] struct Buffer { data: array<u32>; };
+        [[group(0), binding(0)]] var<storage, read_write> buffer: Buffer;
+        [[stage(compute), workgroup_size(1)]] fn main(
+            [[builtin(global_invocation_id)]] id: vec3<u32>) {
+          buffer.data[id.x] = buffer.data[id.x] + 1u;
+        }
+      `,
+    });
+    const module2 = t.device.createShaderModule({
+      code: `
+        [[block]] struct Buffer { data: array<u32>; };
+        [[group(0), binding(0)]] var<storage, read_write> buffer: Buffer;
+        [[stage(compute), workgroup_size(1)]] fn main(
+            [[builtin(global_invocation_id)]] id: vec3<u32>) {
+          buffer.data[id.x] = buffer.data[id.x] + 2u;
+        }
+      `,
+    });
+    const kNumIterations = 250_000;
+    for (let i = 0; i < kNumIterations; ++i) {
+      const module = i % 2 === 0 ? module1 : module2;
+      const pipeline = t.device.createComputePipeline({ compute: { module, entryPoint: 'main' } });
```

Comment:
how different do you want the compute pipelines? Most implementations (I know at least Chrome / Firefox) are going to dedup these to point to exactly the same native object handle. If you really want them to be unique, change the shader contents.

maybe it can be a different test though - it would be a TON of pipelines

Reply:
Ah thanks, that's very good to know. I will perturb the shaders.

Reply:
Settled for 10k pipelines, and I did end up differentiating them only by shader code. Pipeline creation itself ends up being the main bottleneck.